### PR TITLE
3.x.x enhancements

### PR DESCRIPTION
**Pull request by rubensworks**
Several minor enchancements related to the 3.x.x update.

### CODE REVIEW

File: asynciterator.ts [hunk 1]
```diff
@@ -491,7 +491,7 @@ export class AsyncIterator<T> extends EventEmitter {
     After this operation, only read the returned copies instead of the original iterator.
     @returns {module:asynciterator.AsyncIterator} A new iterator that contains all future items of this iterator
   */
-  clone(): AsyncIterator<T> {
+  clone(): ClonedIterator<T> {
```
Comment (RubenVerborgh) [inline asynciterator.ts:494]: Is there a reason we want to give the more specific class? I thought to keep `ClonedIterator` (more or less) internal.

File: asynciterator.ts [hunk 2]
```diff
@@ -1084,7 +1084,7 @@ export class TransformIterator<S, D = S> extends BufferedIterator<D> {
     if (!source || source.done || (item = source.read()) === null)
       done();
     else if (!this._optional)
-      this._transform(item, next);
+      this._transform(item, next, this._push.bind(this));
```
Comment (RubenVerborgh) [inline asynciterator.ts]: probably want a stored `item => this._push(item)` for performance (I can update)
To basically, in the constructor/member variable declaration, `this._pushBound = item => this._push(item)`.

File: asynciterator.ts [hunk 3]
```diff
@@ -1099,7 +1099,7 @@ export class TransformIterator<S, D = S> extends BufferedIterator<D> {
       if (pushedCount === this._pushedCount)
         this._push(item as any as D);
       done();
-    });
+    }, this._push.bind(this));
```
Comment (RubenVerborgh) [inline asynciterator.ts]: idem

File: asynciterator.ts [hunk 4]
```diff
@@ -1252,7 +1253,7 @@ export class SimpleTransformIterator<S, D = S> extends TransformIterator<S, D> {
       // Asynchronously transform the item, and wait for `next` to call back
       else {
         if (!this._optional)
-          this._transform(mappedItem as S, next);
+          this._transform(mappedItem as S, next, this._push.bind(this));
```
Comment (RubenVerborgh) [inline asynciterator.ts]: idem

File: asynciterator.ts [hunk 5]
```diff
@@ -1304,6 +1305,36 @@ export class SimpleTransformIterator<S, D = S> extends TransformIterator<S, D> {
...
+   @param {Array|module:asynciterator.AsyncIterator} [options.prepend] Items to insert before the source items
+   @param {Array|module:asynciterator.AsyncIterator} [options.append]  Items to insert after the source items
+   */
+  constructor(source: AsyncIterator<S>,
+              options?: MultiTransformOptions<S, D> |
+                        MultiTransformOptions<S, D> & ((item: S) => AsyncIterator<D>)) {
+    super(source, options as TransformIteratorOptions<S>);
+
+    // Set transformation steps from the options
+    if (options) {
```
Comment (RubenVerborgh) [inline asynciterator.ts:1332]: This is just new functionality? Or something that broke? (Going to keep it in any case.)

File: asynciterator.ts (hunk 1 above)
Comment (rubensworks) [reply asynciterator.ts:494]: I just checked again, and it was not really needed.

But I would still consider keeping the more specific type though, as ClonedIterator (and TransformIterator) will have some more typed elements, such as `.source`.
But no strong opinion.

File: asynciterator.ts (hunk 2 above)
Comment (rubensworks) [reply asynciterator.ts]: +1 to storing. I wasn't aware that `.bind` is less performant than a lambda.

File: asynciterator.ts (hunk 5 above)
Comment (rubensworks) [reply asynciterator.ts:1332]: Yep, that's just new. I found it annoying that I had to create dedicated classes just for doing small multi-transformations.

File: asynciterator.ts (hunk 2 above)
Comment (RubenVerborgh) [reply asynciterator.ts:1087]: 👍 will keep

File: asynciterator.ts (hunk 1 above)
Comment (RubenVerborgh) [reply asynciterator.ts:494]: 👍 will keep

**Pull request merged**